e2e/nomostest/testutils/workload_identity_test_spec.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutils
+
+import (
+	"fmt"
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/artifactregistry"
+	"kpt.dev/configsync/e2e/nomostest/iam"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/workloadidentity"
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	hubv1 "kpt.dev/configsync/pkg/api/hub/v1"
+	"kpt.dev/configsync/pkg/declared"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/reconcilermanager/controllers"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// WITestSpec defines the test spec for the workload identity test
+type WITestSpec struct {
+	FleetWITest   bool
+	CrossProject  bool
+	SourceRepo    string
+	SourceChart   string
+	SourceVersion string
+	SourceType    v1beta1.SourceType
+	GSAEmail      string
+	RootCommitFn  nomostest.Sha1Func
+}
+
+// TestWorkloadIdentity sets up the test skeleton for both GKE WI and Fleet WI.
+// It has the following requirements:
+// 1. GKE cluster with workload identity enabled.
+// 2. The provided Google service account exists.
+// 3. The cluster is registered to a fleet if testing Fleet WI.
+// 4. IAM permission and IAM policy binding are created.
+func TestWorkloadIdentity(t *testing.T, testSpec WITestSpec) {
```

Comment:
I'm not sure how I feel about this function being moved here, this is moreso the whole testcase implementation rather than a test util.

Reply:
Agreed that this is the whole testcase implementation. 
The original intent is to extract some utility functions into the `testutils` package, such as `registerCluster`, `describeMembership`, and etc.
However, to avoid import cycle, I moved this function along with other utils.

Technically, it is still a test util which relies on actual testcases to provide the arguments.

Do you have any other suggestions?

Reply:
/unhold

Reply:
Perhaps we could write these as a single data driven test with multiple test cases, as we've done elsewhere (for [example](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/testcases/cli_test.go#L674)). I think something like `e2e/testcases/workload_identity_test.go` would make sense to home all of these workload identity tests.

What is the source of the import cycle?

Reply:
If you categorize those tests by the workload identity feature, yes, it makes sense with  `e2e/testcases/workload_identity_test.go`.
The testcases are currently divided by different source types.

Reply:
The import cycle comes from the `ReconcilerPodMissingFWICredsAnnotation` validation function. It is used by both testcases and testutils.

Reply:
I was thinking something like:
```golang
func TestWorkloadIdentity(t *testing.T) {
    testCases := map[string]struct{
        FleetWITest   bool
        CrossProject  bool
        SourceRepo    string
        SourceChart   string
        SourceVersion string
        SourceType    v1beta1.SourceType
        GSAEmail      string
        RootCommitFn  nomostest.Sha1Func
    }{...}
}
```
Primarily to deobfuscate the test logic/assertions from the actual test case.

We could potentially refactor as a subsequent PR, although with the scope of this change I'm concerned about conflicts if we are intending to cherry pick this.

Reply:
That makes sense. I'll update the current PR to avoid cherrypick conflicts.

Reply:
done

Reply:
@karlkfi FYI, CS CI will fail due to this issue and a bunch of bugs will be auto-created and assigned to you as the next onduty.
I will come up with a fix instead of reverting this PR to avoid messy cherrypick history.

Reply:
Ah yeah the `t.Name` must be the same for each of the test cases. We may need to either fix that or revert the change. Sorry about the mixup

Reply:
should be fixed by https://github.com/GoogleContainerTools/kpt-config-sync/pull/1120. I verified with my clusters.